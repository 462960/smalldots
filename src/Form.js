import React, { Component, PropTypes } from 'react'
import isEqual from 'lodash/isEqual'
import get from 'lodash/get'
import cloneDeep from 'lodash/cloneDeep'
import set from 'lodash/set'

export default class Form extends Component {
  static propTypes = { initialValues: PropTypes.object, onSubmit: PropTypes.func }

  static defaultProps = { initialValues: {} }

  state = { values: this.props.initialValues }

  isPristine = () => isEqual(this.state.values, this.props.initialValues)

  isDirty = () => !this.isPristine()

  getValue = path => get(this.state.values, path, '')

  setValue = (path, value) => {
    this.setState({ values: set(cloneDeep(this.state.values), path, value) })
  }

  handleSubmit = event => {
    event.preventDefault()
    if (this.props.onSubmit) {
      this.props.onSubmit(this.state.values)
    }
  }

  reset = () => this.setState({ values: this.props.initialValues })

  render() {
    // eslint-disable-next-line
    const { initialValues, ...rest } = this.props
    return (
      <form onSubmit={this.handleSubmit} {...rest}>
        {this.props.children({
          ...this.state,
          isPristine: this.isPristine,
          isDirty: this.isDirty,
          getValue: this.getValue,
          setValue: this.setValue,
          reset: this.reset
        })}
      </form>
    )
  }
}
