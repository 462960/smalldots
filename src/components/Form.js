import React, { PropTypes } from 'react'

export default function Form(props) {
  const { onSubmit, children, ...otherProps } = props
  return (
    <form {...otherProps} noValidate onSubmit={event => {
      event.preventDefault()
      onSubmit()
    }}>
      {children}
    </form>
  )
}

Form.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired
}
