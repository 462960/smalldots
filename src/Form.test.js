import React from 'react'
import renderer from 'react-test-renderer'
import Form from './Form'

test('renders correctly', () => {
  const tree = renderer.create(
    <Form
      initialValues={{
        foo: 'Foo'
      }}
      validations={{
        foo: [
          value => !value && 'Required'
        ]
      }}
    >
      {() => null}
    </Form>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
