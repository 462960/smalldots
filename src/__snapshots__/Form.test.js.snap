exports[`test renders correctly 1`] = `
<form
  onSubmit={[Function]} />
`;
